integration-test-remote/src/test/java/co/cask/cdap/apps/workflow/WorkflowTest.java

```
@@ -143,8 +143,8 @@ private void testWorkflow(WorkflowManager workflowManager, WikipediaPipelineApp.
           assertWorkflowToken(workflowManager, config, pid, true);
         }
         return true;
-      } catch (Exception e) {
-        // retry on exception
+      } catch (Throwable t) {
+        // retry on throwable, such as AssertionError or NotFoundException
```

We should only retry things that we expect. So if we expect NotFoundException and AssertionError, let's catch those specifically and return false, otherwise let it propagate up.

Otherwise somebody at some point will copy-paste this code as a pattern and retry something for minutes that could have failed immediately.

---

Good point; done.